tests/phpunit/tests/passwordHash.php

```diff
@@ -0,0 +1,45 @@
+<?php
+
+/**
+ * Tests for the PasswordHash external library.
+ *
+ * @covers PasswordHash
+ */
+class Tests_PasswordHash extends WP_UnitTestCase {
+
+	public static function set_up_before_class() {
+		parent::set_up_before_class();
+
+		require_once ABSPATH . WPINC . '/class-phpass.php';
+	}
+
+	/**
+	 * Tests that PasswordHash::gensalt_blowfish() does not throw a deprecation
+	 * notice on PHP 8.1 for "Implicit conversion from float to int loses precision".
+	 *
+	 * @ticket 56340
+	 *
+	 * @covers PasswordHash::gensalt_blowfish
+	 */
+	public function test_gensalt_blowfish_should_not_throw_deprecation_notice_in_php81() {
+		if ( version_compare( PHP_VERSION, '8.1.0', '<' ) ) {
+			$this->markTestSkipped( 'This test should only run on PHP 8.1 or greater.' );
+		}
+
+		try {
+			$this->expectNotToPerformAssertions();
+
+			$hasher = new PasswordHash( 8, true );
+			$hasher->gensalt_blowfish( 'a password string' );
+		} catch ( \Exception $exception ) {
+			$message = $exception->getMessage();
+
+			if ( str_contains( $message, 'Implicit conversion from float' ) ) {
+				$this->fail( 'The deprecation notice was thrown: ' . $message );
+			} else {
+				$this->fail( 'The deprecation notice was not thrown, but another exception was: ' . $message );
+			}
+		}
```

Comment:
To be clear on my reasoning behind not just calling `$hasher->gensalt_blowfish( 'a password string' );` and letting PHPUnit report `E` for this test:

- This implementation means that should the deprecation notice be thrown, it is accurately listed in the "Failures" portion of the report.
- While other exceptions (note: the PHPUnit configuration converts everything to an Exception) will also show a failure `F` rather than reporting an error `E`, the exception's message will still be shown in the failure message. I thought this to be a worthwhile trade-off.

That said, pinging @SergeyBiryukov and @jrfnl for their views on this, as it may be overcomplicating matters with little benefit.

Reply:
IMO, if a PHP notice (deprecation or otherwise) is thrown, the test should error out, not be marked as failed as the _functionality_ of the function hasn't even run completely in that case (PHPUnit < 10 stops as soon as a (converted) exception is seen).

In other words, I'd rather we'd add some tests for the actual function (which would potentially error out without the fix), than to test against specific notices. The tests should also run against all PHP versions. After all, whichever way this is tested, we want the function to run without throwing notices on all supported PHP versions.

As a side-note: might be worth it to do a GitHub  or sourcegraph code search to see if other projects which may be using this library have added tests for it to their project. License-permitting, we could then benefit from a pre-existing set of tests to add for the class.

Reply:
Not sure that I'd be comfortable that I could write the tests to cover the functionality. Unfortunately, I could only find tests that are marked as incomplete with `'This test has not been implemented yet.'` [Ref 1](https://github.com/SimonFormanek/pureosc/blob/06cfa193c5a48e929229394bcf788cc410364d9b/tests/catalog/includes/classes/passwordhashTest.php), [Ref 2](https://github.com/SimonFormanek/pureosc/blob/06cfa193c5a48e929229394bcf788cc410364d9b/tests/catalog/admin/includes/classes/AdminPasswordHashTest.php) 😅

In your opinion, is this PR a no-go? Or could it protect against this notice temporarily until we cover the functionality of `::gensalt_blowfish()`?

Reply:
In my personal opinion: the fix itself is a **go**. I'm not (yet) comfortable with the test(s), but those can be improved/added to during beta/RC, while the fix can no longer go in during beta/RC.

Reply:
If you like, we could do a pair-programming session on these tests once I'm back from my break ?

Reply:
I believe the fix can still go in during Beta as this period is not exclusively for issues introduced during this cycle, but it definitely can't go in during in RC.

A pair-programming session sounds great! 🙂

Reply:
Send me an invite beginning of March for mid-March ?